File: backend/shared-services/e-sign/emailService.ts
/**
 * Email Service for E-Sign Notifications
 * Ported from phase12-esign/backend/email_service.py
 * 
 * Handles all signature-related email notifications:
 * - Signature request emails
 * - Signature completion notifications
 * - Reminder emails
 * - Envelope completion notifications
 */

import nodemailer from 'nodemailer';
import { logger } from '../../src/utils/logger';

// SMTP Configuration
const SMTP_HOST = process.env.SMTP_HOST || 'localhost';
const SMTP_PORT = parseInt(process.env.SMTP_PORT || '1025', 10); // MailHog default
const SMTP_SECURE = process.env.SMTP_SECURE === 'true';
const SMTP_USER = process.env.SMTP_USER;
const SMTP_PASS = process.env.SMTP_PASS;
const EMAIL_FROM = process.env.EMAIL_FROM || 'PropMetrik E-Sign <noreply@propmetrik.com>';

// Create transporter
const transporter = nodemailer.createTransport({
  host: SMTP_HOST,
  port: SMTP_PORT,
  secure: SMTP_SECURE,
  auth: SMTP_USER ? {
    user: SMTP_USER,
    pass: SMTP_PASS,
  } : undefined,
});

// Email Templates
const baseStyles = `
  body { font-family: 'Segoe UI', Arial, sans-serif; line-height: 1.6; color: #333; margin: 0; padding: 0; background: #f4f4f4; }
  .container { max-width: 600px; margin: 0 auto; padding: 20px; }
  .header { background: linear-gradient(135deg, #1a365d 0%, #2563eb 100%); color: white; padding: 30px; text-align: center; border-radius: 10px 10px 0 0; }
  .content { background: #ffffff; padding: 30px; border-radius: 0 0 10px 10px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); }
  .button { display: inline-block; padding: 15px 30px; background: #2563eb; color: white !important; text-decoration: none; border-radius: 8px; margin: 20px 0; font-weight: bold; }
  .button:hover { background: #1d4ed8; }
  .footer { text-align: center; margin-top: 20px; font-size: 12px; color: #666; }
  .document-title { font-size: 18px; font-weight: bold; color: #1a365d; margin: 15px 0; padding: 15px; background: #f8fafc; border-radius: 8px; border-left: 4px solid #2563eb; }
  .warning { color: #dc2626; font-weight: bold; }
  .success { color: #059669; font-weight: bold; }
  .progress-bar { background: #e5e7eb; border-radius: 4px; height: 10px; margin: 15px 0; overflow: hidden; }
  .progress-fill { background: #059669; border-radius: 4px; height: 10px; transition: width 0.3s ease; }
  .status-box { background: #f8fafc; padding: 20px; border-radius: 8px; margin: 20px 0; border: 1px solid #e2e8f0; }
  .security-notice { font-size: 13px; color: #666; background: #fef3c7; padding: 12px; border-radius: 6px; margin-top: 20px; }
`;

export interface SigningRequestEmailParams {
  signerEmail: string;
  signerName: string;
  documentTitle: string;
  creatorName: string;
  signingUrl: string;
  expiresAt?: Date;
  message?: string;
}

export interface SignatureCompletedEmailParams {
  recipientEmail: string;
  recipientName: string;
  documentTitle: string;
  signerName: string;
  completedCount: number;
  totalCount: number;
  viewUrl?: string;
}

export interface ReminderEmailParams {
  signerEmail: string;
  signerName: string;
  documentTitle: string;
  signingUrl: string;
  daysRemaining?: number;
  reminderType: 'initial' | 'follow_up' | 'final_warning';
  creatorName?: string;
}

export interface EnvelopeCompletedEmailParams {
  recipientEmail: string;
  recipientName: string;
  documentTitle: string;
  completedAt: Date;
  downloadUrl?: string;
  certificateUrl?: string;
  signers: { name: string; signedAt: Date }[];
}

export class EmailService {
  /**
   * Send signature request email to a signer
   */
  async sendSigningRequestEmail(params: SigningRequestEmailParams): Promise<boolean> {
    try {
      const expiresText = params.expiresAt
        ? `This request expires on ${params.expiresAt.toLocaleDateString('en-US', { 
            weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' 
          })}.`
        : '';

      const html = `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <style>${baseStyles}</style>
</head>
<body>
  <div class="container">
    <div class="header">
      <h1>📝 Signature Required</h1>
    </div>
    <div class="content">
      <p>Hello <strong>${params.signerName}</strong>,</p>
      
      <p><strong>${params.creatorName}</strong> has requested your signature on:</p>
      
      <div class="document-title">📄 ${params.documentTitle}</div>
      
      ${params.message ? `<p style="font-style: italic; color: #666; padding: 15px; background: #f0f9ff; border-radius: 6px;">"${params.message}"</p>` : ''}
      
      <p>Please review and sign the document by clicking below:</p>
      
      <center>
        <a href="${params.signingUrl}" class="button">Review & Sign Document</a>
      </center>
      
      ${expiresText ? `<p class="warning">⏰ ${expiresText}</p>` : ''}
      
      <div class="security-notice">
        <strong>🔒 Security Notice:</strong> Do not forward this email. 
        The signing link is unique to you and will expire after use.
      </div>
    </div>
    <div class="footer">
      <p>Powered by PropMetrik E-Sign</p>
      <p>© ${new Date().getFullYear()} PropMetrik Ghana Ltd. All rights reserved.</p>
      <p style="font-size: 10px;">Compliant with Ghana Electronic Transactions Act (Act 772)</p>
    </div>
  </div>
</body>
</html>`;

      const text = `
Hello ${params.signerName},

${params.creatorName} has requested your signature on: "${params.documentTitle}"

${params.message ? `Message: "${params.message}"` : ''}

Please click the link below to review and sign the document:
${params.signingUrl}

${expiresText}

Security Notice: Do not forward this email. The signing link is unique to you.

Powered by PropMetrik E-Sign
© ${new Date().getFullYear()} PropMetrik Ghana Ltd. All rights reserved.
`;

      await transporter.sendMail({
        from: EMAIL_FROM,
        to: params.signerEmail,
        subject: `📝 Signature Request: ${params.documentTitle}`,
        text,
        html,
      });

      logger.info('Signing request email sent', {
        to: params.signerEmail,
        document: params.documentTitle,
        from: params.creatorName,
      });
      return true;
    } catch (error) {
      logger.error('Failed to send signing request email', {
        error,
        to: params.signerEmail,
        document: params.documentTitle,
      });
      return false;
    }
  }

  /**
   * Send signature completion notification to creator
   */
  async sendSignatureCompletedEmail(params: SignatureCompletedEmailParams): Promise<boolean> {
    const allComplete = params.completedCount === params.totalCount;
    const progressPercent = Math.round((params.completedCount / params.totalCount) * 100);

    try {
      const html = `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <style>${baseStyles}</style>
</head>
<body>
  <div class="container">
    <div class="header" style="background: ${allComplete ? 'linear-gradient(135deg, #059669 0%, #10b981 100%)' : 'linear-gradient(135deg, #1a365d 0%, #2563eb 100%)'};">
      <h1>${allComplete ? '✅ Document Completed!' : '📝 Signature Received'}</h1>
    </div>
    <div class="content">
      <p>Hello <strong>${params.recipientName}</strong>,</p>
      
      <p><strong>${params.signerName}</strong> has signed:</p>
      
      <div class="status-box">
        <h3 style="margin: 0 0 10px 0;">📄 ${params.documentTitle}</h3>
        <p style="margin: 0; color: #666;">
          Progress: ${params.completedCount} of ${params.totalCount} signatures
        </p>
        <div class="progress-bar">
          <div class="progress-fill" style="width: ${progressPercent}%;"></div>
        </div>
      </div>
      
      ${allComplete ? `
        <p class="success" style="font-size: 16px; text-align: center;">
          🎉 All signers have completed! Your document is ready.
        </p>
        ${params.viewUrl ? `
          <center>
            <a href="${params.viewUrl}" class="button">View & Download Document</a>
          </center>
        ` : ''}
      ` : `
        <p style="color: #666;">
          ⏳ Waiting for ${params.totalCount - params.completedCount} more signature(s).
        </p>
      `}
    </div>
    <div class="footer">
      <p>Powered by PropMetrik E-Sign</p>
      <p>© ${new Date().getFullYear()} PropMetrik Ghana Ltd.</p>
    </div>
  </div>
</body>
</html>`;

      const text = `
Hello ${params.recipientName},

${params.signerName} has signed: "${params.documentTitle}"

Progress: ${params.completedCount} of ${params.totalCount} signatures completed

${allComplete 
  ? '🎉 All signers have completed! Your document is ready.' 
  : `Waiting for ${params.totalCount - params.completedCount} more signature(s).`}

${params.viewUrl ? `View document: ${params.viewUrl}` : ''}

Powered by PropMetrik E-Sign
`;

      await transporter.sendMail({
        from: EMAIL_FROM,
        to: params.recipientEmail,
        subject: allComplete
          ? `✅ Completed: ${params.documentTitle}`
          : `📝 ${params.signerName} signed ${params.documentTitle}`,
        text,
        html,
      });

      logger.info('Signature completed email sent', {
        to: params.recipientEmail,
        document: params.documentTitle,
        allComplete,
      });
      return true;
    } catch (error) {
      logger.error('Failed to send signature completed email', {
        error,
        to: params.recipientEmail,
      });
      return false;
    }
  }

  /**
   * Send reminder email to pending signer
   */
  async sendReminderEmail(params: ReminderEmailParams): Promise<boolean> {
    const urgencyConfig = {
      initial: {
        headerBg: '#f59e0b',
        icon: '⏰',
        title: 'Signature Reminder',
        urgencyText: '',
      },
      follow_up: {
        headerBg: '#f97316',
        icon: '⏰',
        title: 'Signature Reminder',
        urgencyText: 'This is a friendly reminder.',
      },
      final_warning: {
        headerBg: '#dc2626',
        icon: '⚠️',
        title: 'FINAL REMINDER',
        urgencyText: '⚠️ This document will expire soon!',
      },
    };

    const config = urgencyConfig[params.reminderType];

    try {
      const html = `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <style>${baseStyles}</style>
</head>
<body>
  <div class="container">
    <div class="header" style="background: ${config.headerBg};">
      <h2>${config.icon} ${config.title}</h2>
    </div>
    <div class="content">
      <p>Hello <strong>${params.signerName}</strong>,</p>
      
      <p>${config.urgencyText} You have a pending signature request:</p>
      
      <div class="document-title">📄 ${params.documentTitle}</div>
      
      ${params.daysRemaining !== undefined ? `
        <p style="color: #dc2626; font-weight: bold;">
          ⏰ Expires in ${params.daysRemaining} day${params.daysRemaining !== 1 ? 's' : ''}
        </p>
      ` : ''}
      
      ${params.creatorName ? `<p style="color: #666;">Requested by: ${params.creatorName}</p>` : ''}
      
      <center>
        <a href="${params.signingUrl}" class="button">Sign Now</a>
      </center>
      
      <div class="security-notice">
        <strong>🔒 Security Notice:</strong> This link is unique to you. Do not share or forward.
      </div>
    </div>
    <div class="footer">
      <p>Powered by PropMetrik E-Sign</p>
    </div>
  </div>
</body>
</html>`;

      const text = `
Hello ${params.signerName},

${config.urgencyText} You have a pending signature request:

Document: "${params.documentTitle}"
${params.daysRemaining !== undefined ? `Expires in: ${params.daysRemaining} days` : ''}
${params.creatorName ? `Requested by: ${params.creatorName}` : ''}

Sign now: ${params.signingUrl}

Powered by PropMetrik E-Sign
`;

      await transporter.sendMail({
        from: EMAIL_FROM,
        to: params.signerEmail,
        subject: `⏰ ${params.reminderType === 'final_warning' ? 'URGENT: ' : ''}Reminder: ${params.documentTitle} awaits your signature`,
        text,
        html,
      });

      logger.info('Reminder email sent', {
        to: params.signerEmail,
        document: params.documentTitle,
        type: params.reminderType,
      });
      return true;
    } catch (error) {
      logger.error('Failed to send reminder email', {
        error,
        to: params.signerEmail,
        type: params.reminderType,
      });
      return false;
    }
  }

  /**
   * Send envelope completion notification with download links
   */
  async sendEnvelopeCompletedEmail(params: EnvelopeCompletedEmailParams): Promise<boolean> {
    try {
      const signersHtml = params.signers
        .map(s => `<li><strong>${s.name}</strong> - signed ${s.signedAt.toLocaleDateString()}</li>`)
        .join('');

      const html = `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <style>${baseStyles}</style>
</head>
<body>
  <div class="container">
    <div class="header" style="background: linear-gradient(135deg, #059669 0%, #10b981 100%);">
      <h1>🎉 Document Completed!</h1>
    </div>
    <div class="content">
      <p>Hello <strong>${params.recipientName}</strong>,</p>
      
      <p>Great news! All parties have signed the document:</p>
      
      <div class="document-title">📄 ${params.documentTitle}</div>
      
      <div class="status-box">
        <h4 style="margin: 0 0 10px 0;">✅ All Signatures Collected</h4>
        <p style="margin: 0; color: #666;">Completed on ${params.completedAt.toLocaleDateString('en-US', {
          weekday: 'long', year: 'numeric', month: 'long', day: 'numeric'
        })}</p>
        <ul style="margin-top: 15px; padding-left: 20px;">
          ${signersHtml}
        </ul>
      </div>
      
      <center>
        ${params.downloadUrl ? `<a href="${params.downloadUrl}" class="button">📥 Download Signed Document</a>` : ''}
        ${params.certificateUrl ? `<br><a href="${params.certificateUrl}" style="color: #2563eb; text-decoration: underline;">View Certificate of Completion</a>` : ''}
      </center>
      
      <p style="font-size: 13px; color: #666; margin-top: 20px;">
        📋 This document has been digitally signed and timestamped. The Certificate of Completion 
        contains a full audit trail of all signing events.
      </p>
    </div>
    <div class="footer">
      <p>Powered by PropMetrik E-Sign</p>
      <p>© ${new Date().getFullYear()} PropMetrik Ghana Ltd. All rights reserved.</p>
      <p style="font-size: 10px;">Compliant with Ghana Electronic Transactions Act (Act 772)</p>
    </div>
  </div>
</body>
</html>`;

      const text = `
Hello ${params.recipientName},

Great news! All parties have signed the document: "${params.documentTitle}"

Completed on: ${params.completedAt.toLocaleDateString()}

Signers:
${params.signers.map(s => `- ${s.name} (signed ${s.signedAt.toLocaleDateString()})`).join('\n')}

${params.downloadUrl ? `Download: ${params.downloadUrl}` : ''}
${params.certificateUrl ? `Certificate: ${params.certificateUrl}` : ''}

Powered by PropMetrik E-Sign
`;

      await transporter.sendMail({
        from: EMAIL_FROM,
        to: params.recipientEmail,
        subject: `🎉 Completed: ${params.documentTitle}`,
        text,
        html,
      });

      logger.info('Envelope completed email sent', {
        to: params.recipientEmail,
        document: params.documentTitle,
      });
      return true;
    } catch (error) {
      logger.error('Failed to send envelope completed email', {
        error,
        to: params.recipientEmail,
      });
      return false;
    }
  }

  /**
   * Verify SMTP connection is working
   */
  async verifyConnection(): Promise<boolean> {
    try {
      await transporter.verify();
      logger.info('SMTP connection verified successfully');
      return true;
    } catch (error) {
      logger.error('SMTP connection verification failed', { error });
      return false;
    }
  }
}

export const emailService = new EmailService();

File: phase12-esign/backend/email_service.py
"""
Email Service for Signature Request Notifications
Sends emails via SMTP (MailHog in development)
"""

import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from typing import List
from config import settings


def send_signature_request_email(
    signer_email: str,
    signer_name: str,
    document_title: str,
    creator_name: str,
    signing_url: str,
    expires_at: str = None
) -> bool:
    """
    Send signature request email to a signer
    
    Args:
        signer_email: Email address of the signer
        signer_name: Full name of the signer
        document_title: Title of the document to sign
        creator_name: Name of person who created the request
        signing_url: URL for the signer to access and sign the document
        expires_at: Optional expiration date string
    
    Returns:
        bool: True if email sent successfully, False otherwise
    """
    try:
        # Create message
        message = MIMEMultipart("alternative")
        message["From"] = f"Cedyn E-Signature <noreply@cedynhq.com>"
        message["To"] = signer_email
        message["Subject"] = f"Signature Request: {document_title}"
        
        # Create plain text version
        text_content = f"""
Hello {signer_name},

{creator_name} has requested your signature on the document: "{document_title}"

Please click the link below to review and sign the document:
{signing_url}

{f"This request expires on: {expires_at}" if expires_at else ""}

If you have any questions, please contact {creator_name}.

Thank you,
Cedyn E-Signature Platform
"""
        
        # Create HTML version
        html_content = f"""
<!DOCTYPE html>
<html>
<head>
    <style>
        body {{ font-family: Arial, sans-serif; line-height: 1.6; color: #333; }}
        .container {{ max-width: 600px; margin: 0 auto; padding: 20px; }}
        .header {{ background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 30px; text-align: center; border-radius: 10px 10px 0 0; }}
        .content {{ background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px; }}
        .button {{ display: inline-block; padding: 15px 30px; background: #667eea; color: white; text-decoration: none; border-radius: 5px; margin: 20px 0; }}
        .footer {{ text-align: center; margin-top: 20px; font-size: 12px; color: #666; }}
        .document-title {{ font-size: 18px; font-weight: bold; color: #667eea; margin: 15px 0; }}
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>🖊️ Signature Request</h1>
        </div>
        <div class="content">
            <p>Hello <strong>{signer_name}</strong>,</p>
            
            <p><strong>{creator_name}</strong> has requested your signature on the following document:</p>
            
            <div class="document-title">📄 {document_title}</div>
            
            <p>Please review and sign the document by clicking the button below:</p>
            
            <center>
                <a href="{signing_url}" class="button">Review & Sign Document</a>
            </center>
            
            {f'<p style="color: #e74c3c;"><strong>⏰ Expires:</strong> {expires_at}</p>' if expires_at else ''}
            
            <p>If you have any questions about this document, please contact {creator_name}.</p>
            
            <p>Thank you for using Cedyn E-Signature Platform!</p>
        </div>
        <div class="footer">
            <p>This is an automated message from Cedyn E-Signature Platform</p>
            <p>© 2025 Cedyn Group. All rights reserved.</p>
        </div>
    </div>
</body>
</html>
"""
        
        # Attach both versions
        part1 = MIMEText(text_content, "plain")
        part2 = MIMEText(html_content, "html")
        message.attach(part1)
        message.attach(part2)
        
        # Send email
        with smtplib.SMTP(settings.SMTP_HOST, settings.SMTP_PORT) as server:
            if settings.SMTP_USE_TLS:
                server.starttls()
            if settings.SMTP_USERNAME and settings.SMTP_PASSWORD:
                server.login(settings.SMTP_USERNAME, settings.SMTP_PASSWORD)
            
            server.send_message(message)
        
        print(f"✅ Signature request email sent to {signer_email}")
        return True
        
    except Exception as e:
        print(f"❌ Failed to send email to {signer_email}: {e}")
        return False


def send_signature_completed_email(
    creator_email: str,
    creator_name: str,
    document_title: str,
    signer_name: str,
    completed_count: int,
    total_count: int
) -> bool:
    """
    Notify document creator when a signer completes their signature
    
    Args:
        creator_email: Email of the request creator
        creator_name: Name of the creator
        document_title: Document title
        signer_name: Name of person who just signed
        completed_count: Number of signers who have completed
        total_count: Total number of signers
    
    Returns:
        bool: True if email sent successfully
    """
    try:
        is_fully_signed = completed_count == total_count
        subject = f"{'✅ All Signatures Completed' if is_fully_signed else '✍️ Signature Received'}: {document_title}"
        
        message = MIMEMultipart("alternative")
        message["From"] = f"Cedyn E-Signature <noreply@cedynhq.com>"
        message["To"] = creator_email
        message["Subject"] = subject
        
        text_content = f"""
Hello {creator_name},

{signer_name} has signed the document: "{document_title}"

Progress: {completed_count} of {total_count} signatures completed

{"🎉 All signatures have been collected! The document is now fully executed." if is_fully_signed else f"Waiting for {total_count - completed_count} more signature(s)."}

Thank you for using Cedyn E-Signature Platform!
"""
        
        html_content = f"""
<!DOCTYPE html>
<html>
<head>
    <style>
        body {{ font-family: Arial, sans-serif; line-height: 1.6; color: #333; }}
        .container {{ max-width: 600px; margin: 0 auto; padding: 20px; }}
        .header {{ background: {"#27ae60" if is_fully_signed else "#3498db"}; color: white; padding: 30px; text-align: center; border-radius: 10px 10px 0 0; }}
        .content {{ background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px; }}
        .progress {{ background: #e0e0e0; height: 20px; border-radius: 10px; margin: 20px 0; }}
        .progress-bar {{ background: {"#27ae60" if is_fully_signed else "#3498db"}; height: 100%; border-radius: 10px; width: {(completed_count/total_count)*100}%; }}
        .footer {{ text-align: center; margin-top: 20px; font-size: 12px; color: #666; }}
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>{"🎉 All Signatures Completed!" if is_fully_signed else "✍️ Signature Received"}</h1>
        </div>
        <div class="content">
            <p>Hello <strong>{creator_name}</strong>,</p>
            
            <p><strong>{signer_name}</strong> has successfully signed:</p>
            <p style="font-size: 18px; font-weight: bold; color: #667eea;">📄 {document_title}</p>
            
            <p><strong>Progress:</strong></p>
            <div class="progress">
                <div class="progress-bar"></div>
            </div>
            <p style="text-align: center;"><strong>{completed_count} of {total_count}</strong> signatures completed</p>
            
            {f'<p style="color: #27ae60; font-size: 18px; text-align: center;">🎉 <strong>All signatures collected!</strong><br>The document is now fully executed.</p>' if is_fully_signed else f'<p>Waiting for <strong>{total_count - completed_count}</strong> more signature(s).</p>'}
            
            <p>Thank you for using Cedyn E-Signature Platform!</p>
        </div>
        <div class="footer">
            <p>© 2025 Cedyn Group. All rights reserved.</p>
        </div>
    </div>
</body>
</html>
"""
        
        part1 = MIMEText(text_content, "plain")
        part2 = MIMEText(html_content, "html")
        message.attach(part1)
        message.attach(part2)
        
        with smtplib.SMTP(settings.SMTP_HOST, settings.SMTP_PORT) as server:
            if settings.SMTP_USE_TLS:
                server.starttls()
            if settings.SMTP_USERNAME and settings.SMTP_PASSWORD:
                server.login(settings.SMTP_USERNAME, settings.SMTP_PASSWORD)
            
            server.send_message(message)
        
        print(f"✅ Completion notification sent to {creator_email}")
        return True
        
    except Exception as e:
        print(f"❌ Failed to send completion email: {e}")
        return False
